pkg/inventory/model/client.go

```
@@ -0,0 +1,281 @@
+	}
+	lv := reflect.ValueOf(list)
+	lt := reflect.TypeOf(list)
+	switch lt.Kind() {
+	case reflect.Ptr:
+		lv = lv.Elem()
+		lt = lt.Elem()
+	default:
+		return nil
+	}
+	switch lv.Kind() {
+	case reflect.Slice:
+		l, err := Table{r.db}.List(model, options)
+		if err != nil {
+			Log.Trace(err)
+			return err
+		}
+		concrete := reflect.MakeSlice(lv.Type(), 0, 0)
+		for i := 0; i < len(l); i++ {
+			m := reflect.ValueOf(l[i]).Elem()
+			concrete = reflect.Append(concrete, m)
+		}
+		lv.Set(concrete)
+	}
+
+	return nil
+}
+
+//
+// Begin a transaction.
+// Example:
+//   tx, _ := client.Begin()
+//   defer tx.Rollback()
+//   client.Insert(model)
+//   client.Insert(model)
+//   tx.Commit()
+func (r *Client) Begin() (*Tx, error) {
+	r.Lock()
+	defer r.Unlock()
+	r.mutex.Lock()
+	tx, err := r.db.Begin()
+	if err != nil {
+		return nil, err
+	}
+	r.tx = tx
+	return &Tx{client: r, ref: tx}, nil
+}
+
+//
+// Insert the model.
+func (r *Client) Insert(model Model) error {
+	r.Lock()
+	defer r.Unlock()
+	model.SetPk()
+	table := Table{}
+	if r.tx == nil {
+		r.mutex.Lock()
+		defer r.mutex.Unlock()
+		table.Db = r.db
+	} else {
+		table.Db = r.tx
+	}
+
+	return table.Insert(model)
+}
+
+//
+// Update the model.
+func (r *Client) Update(model Model) error {
+	r.Lock()
+	defer r.Unlock()
+	model.SetPk()
+	table := Table{}
+	if r.tx == nil {
+		r.mutex.Lock()
+		defer r.mutex.Unlock()
+		table.Db = r.db
+	} else {
+		table.Db = r.tx
+	}
+
+	return table.Update(model)
+}
+
+//
+// Delete the model.
+func (r *Client) Delete(model Model) error {
+	r.Lock()
+	defer r.Unlock()
+	model.SetPk()
+	table := Table{}
+	if r.tx == nil {
+		r.mutex.Lock()
+		defer r.mutex.Unlock()
+		table.Db = r.db
+	} else {
+		table.Db = r.tx
+	}
+
+	return table.Delete(model)
+}
+
+//
+// Commit a transaction.
+// This MUST be preceeded by Begin() which returns
+// the `tx` transaction token.
+func (r *Client) commit(tx *Tx) error {
+	r.Lock()
+	defer r.Unlock()
+	if r.tx == nil || r.tx != tx.ref {
```

I wonder if this branch ought to panic or return an error rather than silently returning.

---

Yes, I think so.  Updated to return an error.